quinn-proto/src/token.rs

```diff
@@ -19,79 +19,79 @@ use crate::{
 // - AEAD nonce is always set to 0
 // in other words, for each ticket, use different key derived from random using HKDF
 
-pub fn generate<K>(
-    key: &K,
-    address: &SocketAddr,
-    src_cid: &ConnectionId,
-    dst_cid: &ConnectionId,
-    issued: SystemTime,
-) -> Vec<u8>
-where
-    K: HmacKey,
-{
-    let mut buf = Vec::new();
+pub struct Token {
+    pub src_cid: ConnectionId,
+    pub dst_cid: ConnectionId,
+    pub issued: SystemTime,
+}
 
-    buf.write(src_cid.len() as u8);
-    buf.put_slice(src_cid);
+impl Token {
+    pub fn encode(&self, key: &impl HmacKey, address: &SocketAddr) -> Vec<u8> {
+        let mut buf = Vec::new();
 
-    buf.write(dst_cid.len() as u8);
-    buf.put_slice(dst_cid);
+        buf.write(self.src_cid.len() as u8);
+        buf.put_slice(&self.src_cid);
 
-    buf.write::<u64>(
-        issued
-            .duration_since(UNIX_EPOCH)
-            .map(|x| x.as_secs())
-            .unwrap_or(0),
-    );
-    let signature_pos = buf.len();
-    match address.ip() {
-        IpAddr::V4(x) => buf.put_slice(&x.octets()),
-        IpAddr::V6(x) => buf.put_slice(&x.octets()),
-    }
-    buf.write(address.port());
-    let signature = key.sign(&buf);
-    // No reason to actually encode the IP in the token, since we always have the remote addr for an incoming packet.
-    buf.truncate(signature_pos);
-    buf.extend_from_slice(signature.as_ref());
-    buf
-}
+        buf.write(self.dst_cid.len() as u8);
+        buf.put_slice(&self.dst_cid);
 
-pub fn check<K>(
-    key: &K,
-    address: &SocketAddr,
-    data: &[u8],
-) -> Option<(ConnectionId, ConnectionId, SystemTime)>
-where
-    K: HmacKey,
-{
-    let mut reader = io::Cursor::new(data);
-    let src_cid_len = reader.get::<u8>().ok()? as usize;
-    if src_cid_len > reader.remaining() || src_cid_len > MAX_CID_SIZE {
-        return None;
+        buf.write::<u64>(
+            self.issued
+                .duration_since(UNIX_EPOCH)
+                .map(|x| x.as_secs())
+                .unwrap_or(0),
+        );
+        let signature_pos = buf.len();
+        match address.ip() {
+            IpAddr::V4(x) => buf.put_slice(&x.octets()),
+            IpAddr::V6(x) => buf.put_slice(&x.octets()),
+        }
+        buf.write(address.port());
+        let signature = key.sign(&buf);
+        // No reason to actually encode the IP in the token, since we always have the remote addr for an incoming packet.
+        buf.truncate(signature_pos);
+        buf.extend_from_slice(signature.as_ref());
+        buf
     }
-    let src_cid = ConnectionId::new(&reader.bytes()[..src_cid_len]);
-    reader.advance(src_cid_len);
 
-    let dst_cid_len = reader.get::<u8>().ok()? as usize;
-    if dst_cid_len > reader.remaining() || dst_cid_len > MAX_CID_SIZE {
-        return None;
-    }
-    let dst_cid = ConnectionId::new(&reader.bytes()[..dst_cid_len]);
-    reader.advance(dst_cid_len);
+    pub fn decode_and_validate(
```

Comment:
This name seems a little unwieldy, I'd probably just call it `from_bytes()` or `from_slice()`. Also maybe return `Result<Self, ()>` rather than `Option`, especially if returning `None` means it gets "rejected"?